crates/wasi-http/src/body.rs

```diff
@@ -423,42 +454,115 @@ impl HostOutgoingBody {
             finish_receiver: Some(finish_receiver),
         }
         .boxed();
+
+        // TODO: this capacity constant is arbitrary, and should be configurable
+        let output_stream = BodyWriteStream::new(1024 * 1024, body_sender, written.clone());
+
         (
             Self {
-                // TODO: this capacity constant is arbitrary, and should be configurable
-                body_output_stream: Some(Box::new(BodyWriteStream::new(1024 * 1024, body_sender))),
+                body_output_stream: Some(Box::new(output_stream)),
+                written,
                 finish_sender: Some(finish_sender),
             },
             body_impl,
         )
     }
+
+    pub fn finish(mut self, ts: Option<FieldMap>) -> Result<(), types::ErrorCode> {
+        // Make sure that the output stream has been dropped, so that the BodyImpl poll function
+        // will immediately pick up the finish sender.
+        drop(self.body_output_stream);
+
+        let sender = self
+            .finish_sender
+            .take()
+            .expect("outgoing-body trailer_sender consumed by a non-owning function");
+
+        // TODO: does sending guarantee that the message was received, or could it be queued to
+        // receive?
```

Comment:
I'm not 100% sure as to the context of this comment but sending on a oneshot does not guarantee the message is received, it's only queued up at that time to possible be received in the future.

Reply:
I was wondering if we would potentially block on the oneshot send if there was nothing already waiting to receive it. It sounds like that's not the case, so I'll remove the TODO. Thanks!